**Contents**

1. Exploring the Indian Crop Production Statistics Data
2. Conducting EDA
3. Checking for mixed-type columns, duplicates and missing values
3. Cleaning up duplicates and missing values

In [1]:
import numpy as np
import pandas as pd

In [2]:
df = pd.read_csv('/content/Crop Production Statistics - India.csv')

1. Exploring the Indian Crop Production Statistics Data

In [3]:
df.head()

State District       Crop  Crop_Year       Season  \
0  Andaman and Nicobar Island  NICOBARS  Arecanut       2007  Kharif        
1  Andaman and Nicobar Island  NICOBARS  Arecanut       2007  Rabi          
2  Andaman and Nicobar Island  NICOBARS  Arecanut       2008  Autumn        
3  Andaman and Nicobar Island  NICOBARS  Arecanut       2008  Summer        
4  Andaman and Nicobar Island  NICOBARS  Arecanut       2009  Autumn        

    Area   Production  Yield  
0  2439.6      3415.0   1.40  
1  1626.4      2277.0   1.40  
2  4147.0      3060.0   0.74  
3  4147.0      2660.0   0.64  
4  4153.0      3120.0   0.75

In [4]:
#Exploring the number of rows and columns

df.shape

(345336, 8)

3. Checking for mixed-type columns, duplicates and missing values

In [5]:
#Checking for null values 

df.isnull().sum()

State            0
District         0
Crop             9
Crop_Year        0
Season           0
Area             0
Production    4948
Yield            0
dtype: int64

In [6]:
df[df['Production']==0].shape

(1466, 8)

In [7]:
df[df['Production']==np.nan].shape

(0, 8)

In [8]:
df[df['Yield']==0].shape

(6096, 8)

In [9]:
df[df['Yield']==np.nan].shape

(0, 8)

There are 1466 0s in the Production column and 6096 0s in the Yield column. We will be dealing with the yield column first, since yield = production/ Area. We will first fill the missing values in Yield column with Production/Area. Then we will fill in the missing values in the Production column with Area * Yield. 

In [10]:
#Checking for mixed-type data in df dataframe

for col in df.columns.tolist():
  weird = (df[[col]].applymap(type) != df[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df[weird]) > 0:
    print (col)

Crop


The Crop name column is mixed type. 

In [11]:
#Checking for duplicates in df dataframe

df_dups = df[df.duplicated()]

In [12]:
df_dups

Empty DataFrame
Columns: [State, District , Crop, Crop_Year, Season, Area , Production, Yield]
Index: []

No duplicates

2. Conducting EDA

In [13]:
#Conducting EDA 

df.describe()

Crop_Year         Area     Production          Yield
count  345336.000000  3.453360e+05  3.403880e+05  345336.000000
mean     2008.887512  1.167147e+04  9.584726e+05      79.423135
std         6.564361  4.584079e+04  2.153068e+07     916.678396
min      1997.000000  4.000000e-03  0.000000e+00       0.000000
25%      2003.000000  7.400000e+01  8.700000e+01       0.550000
50%      2009.000000  5.320000e+02  7.170000e+02       1.000000
75%      2015.000000  4.112000e+03  7.182000e+03       2.470000
max      2020.000000  8.580100e+06  1.597800e+09   43958.330000

4. Cleaning up duplicates and missing values

In [14]:
#Filling the 0 values in the Yield column with Production/Area

df['Yield'] = df.apply(lambda x: x['Production']/x['Area '] if x['Yield'] == 0 
                        else x['Yield'], axis = 1)


In [15]:
df.isnull().sum()

State            0
District         0
Crop             9
Crop_Year        0
Season           0
Area             0
Production    4948
Yield         4948
dtype: int64

In [16]:
df[df['Yield']==0].shape

(1032, 8)

In [17]:
df[df['Production']==0].shape

(1466, 8)

In [18]:
df[df['Yield']== np.nan].shape

(0, 8)

In [19]:
df[df['Production']==np.nan].shape

(0, 8)

In [20]:
#Filling the 0 values in the Production column with Area*Yield

df['Production'] = df.apply(lambda x: x['Area ']*x['Yield'] if x['Production'] == 0 
                        else x['Production'], axis = 1)

In [21]:
df[df['Production']==0].shape

(1032, 8)

In [22]:
df[df['Yield']==0].shape

(1032, 8)

Now the same number of 0 values in the Production and Yield columns are understandable. Some land produced 0 metric tonnes of crops and resulted in 0 yields. 

In [23]:
df.isnull().sum()

State            0
District         0
Crop             9
Crop_Year        0
Season           0
Area             0
Production    4948
Yield         4948
dtype: int64

In [25]:
#Filling the null values in the Yield column with Production/Area

df['Yield'] = df.apply(lambda x: x['Production']/x['Area '] if np.isnan(x['Yield'])
                        else x['Yield'], axis = 1)


In [26]:
#Filling the null values in the Production column with Area*Yield

df['Production'] = df.apply(lambda x: x['Area ']*x['Yield'] if np.isnan(x['Production']) 
                        else x['Production'], axis = 1)

In [27]:
df.isnull().sum()

State            0
District         0
Crop             9
Crop_Year        0
Season           0
Area             0
Production    4948
Yield         4948
dtype: int64

4b. Fixing the mixed-type column



In [28]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 345336 entries, 0 to 345335
Data columns (total 8 columns):
 #   Column      Non-Null Count   Dtype  
---  ------      --------------   -----  
 0   State       345336 non-null  object 
 1   District    345336 non-null  object 
 2   Crop        345327 non-null  object 
 3   Crop_Year   345336 non-null  int64  
 4   Season      345336 non-null  object 
 5   Area        345336 non-null  float64
 6   Production  340388 non-null  float64
 7   Yield       340388 non-null  float64
dtypes: float64(3), int64(1), object(4)
memory usage: 21.1+ MB


Here, crop is an object type variable, which means it's a string. Hence we won't need tofurther convert it to a string, since the name of crops is supposed to be string

In [29]:
#Renaming the 'Area ' column

df.rename(columns = {'Area ' : 'Area'}, inplace = True)

In [30]:
df['Area']

0         2439.6
1         1626.4
2         4147.0
3         4147.0
4         4153.0
           ...  
345331     855.0
345332    1366.0
345333    1052.0
345334     833.0
345335     516.0
Name: Area, Length: 345336, dtype: float64

In [31]:
#Downloading the Preprocessed file

from google.colab import files
df.to_csv('India_crops_preprocessed.csv') 
files.download('India_crops_preprocessed.csv')

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>